pyproject.toml

```diff
@@ -0,0 +1,13 @@
+[tool.ruff]
```

Comment:
Please, no pyproject.toml in SAGE_ROOT. 
SAGE_ROOT is not a Python package.

Reply:
It applies to all python code (e.g. also the one in pkgs), so should be in the root. It's also quite common that the pyproject config is in the root, but the actual code in a subfolder. eg https://github.com/scipy/scipy/blob/main/pyproject.toml or https://github.com/pandas-dev/pandas/blob/main/pyproject.toml

Reply:
> It applies to all python code (e.g. also the one in pkgs), so should be in the root.

No, that's not how the Sage repository is structured. Again, SAGE_ROOT is not the source directory of a Python package (distribution)... which is why there is no setup.py, no pyproject.toml, no setup.cfg, etc.

Reply:
How is that different from scipy?

Reply:
The source directory of the Python distribution package scipy (per PEP 518, PEP 621, the location of the `pyproject.toml` file) is the root of its repository https://github.com/scipy/scipy/tree/main

The source directory of the Python distribution package sagemath-standard is not the root of the repository https://github.com/sagemath/sage (= SAGE_ROOT), but rather `src/`. As I know that you know, this is where you find `src/pyproject.toml`, `src/setup.py`, etc.

Reply:
I didn't changed anything about a distribution package. See it similar to the tox.ini file.

Do you also have a suggestion on how to apply config otherwise for tools that look at pyproject and that should apply to all python code?

Reply:
> Do you also have a suggestion on how to apply config otherwise for tools that look at pyproject and that should apply to all python code?

By design, all source code of the Sage library is in the monolithic source directory `SAGE_ROOT/src`. So there is no need for tools to look at the source files of the modularized distribution packages (sagemath-environment etc.)
Some special distribution packages (currently only pkgs/sage-sws2rst, pkgs/sage-conf*) do have their own source files, so tool configuration could go into pkgs/sage-sws2rst/pyproject.toml etc.

However, Ruff also supports configuration using `ruff.toml` https://docs.astral.sh/ruff/configuration/#using-rufftoml, so this would be a good way to put the tool configuration that is "global to the repository" there

Reply:
> Some special distribution packages (currently only pkgs/sage-sws2rst, pkgs/sage-conf*) do have their own source files, so tool configuration could go into pkgs/sage-sws2rst/pyproject.toml etc.

That doesn't scale. I'll not add ruff config to each pyproject toml file.

> However, Ruff also supports configuration using `ruff.toml` https://docs.astral.sh/ruff/configuration/#using-rufftoml, so this would be a good way to put the tool configuration that is "global to the repository" there

I think there is a reason why we put all the config's in tox ini so far instead of having 20 config files for all tools.

Not everything that has a pyproject file is a "distribution package". Especially not if it only contains tool config...

Reply:
> > Some special distribution packages (currently only pkgs/sage-sws2rst, pkgs/sage-conf*) do have their own source files, so tool configuration could go into pkgs/sage-sws2rst/pyproject.toml etc.
> 
> That doesn't scale.

It does not have to scale. As I just explained to you, there is a fixed short list of packages.

Reply:
> Not everything that has a pyproject file is a "distribution package".

That's the very idea of pyproject.toml. It marks the top of a Python distribution package (= project). That's what "py" and "project" refer to in the name of this file.

Reply:
> > > Some special distribution packages (currently only pkgs/sage-sws2rst, pkgs/sage-conf*) do have their own source files, so tool configuration could go into pkgs/sage-sws2rst/pyproject.toml etc.
> > 
> > 
> > That doesn't scale.
> 
> It does not have to scale. As I just explained to you, there is a fixed short list of packages.

Copying the config 4 times is already too much for my taste. Thanks for the suggestion, but I'll not implement it.

Reply:
So do ruff.toml, as I explained.

Reply:
> So do ruff.toml, as I explained.

I've tried that but couldn't make it work reliable. For some files it just doesn't seem to work (although, I agree, that the docs claim that ruff.toml and pyproject.toml are equivalent). As a starter, you need to put the config in the root so that it applies to all python files in all directories. Moreover, it sometimes just didn't apply the config when reformatting a file - not sure if it's a problem with ruff or vscode, or something else. My guess is that the somewhat strange directory layout with a pyproject.toml in src is resulting in hick ups.

Since the config in the root pyproject.toml worked exellently for me, I stopped investigating. The added pyproject.toml in the root doesn't cause any other issues, is helpful in other contexts as well (eg #37446) and `pip install .` will work correctly with #36524.

Sorry, should have let you know before you invested the time to prepare https://github.com/sagemath/sage/pull/37452.

Reply:
> > Not everything that has a pyproject file is a "distribution package".
> 
> That's the very idea of pyproject.toml. It marks the top of a Python distribution package (= project). That's what "py" and "project" refer to in the name of this file.

That's a very narrow viewpoint. Multiple tools support non-distribution/non-package scenarios via `pyproject.toml`, see eg https://python-poetry.org/docs/basic-usage#operating-modes. In https://discuss.python.org/t/projects-that-arent-meant-to-generate-a-wheel-and-pyproject-toml/29684/5 a Cpython core developer addressed the concern that pyproject.toml indicates that code which will be installed as a distribution as follows:
 > [...] storing configuration for black/ruff etc is not a reasonable indicator that code contained in that directory is intended to be installed as a distribution. The project section, yes, is currently only usable with distribution; but it’s certainly not an indicator without that. [...] That isn’t the case and it’s not a packaging-exclusive marker file today; and it hasn’t been since its introduction.

According to this viewpoint, since I don't add a `project` section, the `pyproject.toml` in the root does not function as a indicator for a distribution package.